File: Arrays/matrix_zeroify.py
def pretty_print(arr):
    for i in arr:
        print(i)


def zeroify(arr):
    zero_rows = set()
    zero_cols = set()
    for row_index, row in enumerate(arr):
        for col_index, val in enumerate(row):
            if val == 0:
                zero_rows.add(row_index)
                zero_cols.add(col_index)

    row_length = len(arr)
    col_length = len(arr[0])


    for i in range(0, row_length):
        for j in range(0, col_length):
            if i in zero_rows or j in zero_cols:
                arr[i][j] = 0

    
    
    
def main():
    input_arr = [[1, 1, 1],
                [1, 0, 1],
                [0, 1, 1]]
    print("Input array: ")
    pretty_print(input_arr)


    zeroify(input_arr)

    print("Zeroified array: ")
    pretty_print(input_arr)


if __name__ == '__main__':
    main()

File: Arrays/matrix_zeroify.cpp
#include <iostream>
#include <set>

void display_arr(int arr[3][3]) {
    for (int i=0; i<3; i++) {
        for (int j=0; j<3; j++) {
            printf("%d ", arr[i][j]);
        }
        printf("\n");
    }
}

void zeroify(int arr[3][3]) {
    std::set<int> zero_rows;
    std::set<int> zero_cols;

    for (int i=0; i<3; i++) {
        for (int j=0; j<3; j++) {
            if (arr[i][j] == 0) {
                zero_rows.insert(i);
                zero_cols.insert(j);
            }
        }
    }

    for (int i=0; i<3; i++) {
        for (int j=0; j<3; j++) {
            if (zero_rows.find(i) != zero_rows.end() ||
                zero_cols.find(j) != zero_cols.end())
                arr[i][j] = 0;
        }
    }

    
}

int main() {
    int arr[3][3] = {{1, 1, 1},
                    {1, 0, 1},
                    {0, 1, 1}};

    printf("Input array is: \n");
    display_arr(arr);

    zeroify(arr);
    printf("Zeroified array is: \n");
    display_arr(arr);

    return 0;
}